linked_list/linked_list.py

```diff
@@ -86,32 +163,84 @@ def visit(self):
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)?
+    # Space Complexity: O(1)?
     def reverse(self):
-        pass
+        prev = None
+        current = self.head
+        while current is not None:
+            next = current.next
+            current.next = prev
+            prev = current
+            current = next
+        self.head = prev
   
     ## Advanced/ Exercises
     # returns the value at the middle element in the singly linked list
     # Time Complexity: ?
     # Space Complexity: ?
+    # find middle value of a singly linked list
     def find_middle_value(self):
-        pass
+        if self.head == None:
+            return None
+
+        current = self.head
+        count = 0
+        while current:
+            count += 1
+            current = current.next

```

Comment:
hmm where have we seen this already? Maybe we can call `length` to get the number of nodes